### PR TITLE
fix applyconfiguration-gen

### PR DESCRIPTION
**Pull request by xiaoyawei**
#### What type of PR is this?

/kind bug

#### What this PR does / why we need it:

There's bug in `applyconfiguration-gen` that does not properly parse the `external-applyconfiguration` param; this PR fixes it. It also includes `OwnerReference` in the default value since it is also pretty commonly needed.

#### Which issue(s) this PR fixes:

I just came across this bug and didn't file an issue before I found its root cause. Let me know if I need to create an issue along the way as well

#### Does this PR introduce a user-facing change?

NO

```release-note
NONE
```

#### Additional documentation e.g., KEPs (Kubernetes Enhancement Proposals), usage docs, etc.:

```docs

```

### CODE REVIEW

**Comment by xiaoyawei** [comment]
hi @wojtek-t, would you mind taking a look when you have time? thanks!

**Comment by xiaoyawei** [comment]
hi @wojtek-t, would you mind taking a look when you have time? thanks!

**Comment by wojtek-t** [comment]
Sorry for delay. I think I don't understand this bug - can you describe what exactly the problem currently is?

**Comment by xiaoyawei** [comment]
hi @wojtek-t , sure thing. I came up with this PR primarily because I went through a couple of issues when I was trying to use `applyconfiguration-gen` to generate codes for `apply` method in a CRD client. The issues include 

1. In the generated apply configurations, there is a `ObjectMetaApplyConfiguration` field; but in definition of `ObjectMetaApplyConfiguration`, there the `OwnerReference` field has type `apis/meta/v1.OwnerReference`, instead of `applyconfigurations/meta/v1.OwnerReferenceApplyConfiguration`.
2. the `applyconfigurations-gen` also takes locations of external generated apply configurations as mentioned in https://github.com/kubernetes/kubernetes/pull/110802/files#diff-aac88022374b6fac8b3dd39987d75ccd1a0e4ed45607d6e32fb888d6104320f0L32-L35; currently user is supposed to use `#` as delimiter to separate the package and struct name (and we can't use `.` as delimiter as the package could be like `k8s.io/apimachinery/pkg/apis/meta/v1` which already contains `.`), but the corresponding argument parser didn't use `#` but still uses `.`;

These are the 2 main issues that this PR is trying to fix; let me know if you need more explainations

File: staging/src/k8s.io/code-generator/cmd/applyconfiguration-gen/args/args.go [hunk 1]
```diff
@@ -66,7 +67,7 @@ func NewDefaults() (*args.GeneratorArgs, *CustomArgs) {
 func (ca *CustomArgs) AddFlags(fs *pflag.FlagSet, inputBase string) {
 	pflag.Var(NewExternalApplyConfigurationValue(&ca.ExternalApplyConfigurations, nil), "external-applyconfigurations",
 		"list of comma separated external apply configurations locations in <type-package>.<type-name>:<applyconfiguration-package> form."+
-			"For example: k8s.io/api/apps/v1.Deployment:k8s.io/client-go/applyconfigurations/apps/v1")
+			"For example: k8s.io/api/apps/v1#Deployment:k8s.io/client-go/applyconfigurations/apps/v1")
```
Comment (apelisse) [inline staging/src/k8s.io/code-generator/cmd/applyconfiguration-gen/args/args.go:70]: > "list of comma separated external apply configurations locations in \<type-package\>.\<type-name\>:<applyconfiguration-package> form.

The flag description is incorrect if you change that

File: staging/src/k8s.io/code-generator/cmd/applyconfiguration-gen/args/externaltypes.go [hunk 2]
```diff
@@ -114,7 +114,7 @@ func parseExternalMapping(mapping string) (typ types.Name, pkg string, err error
 	}
 	packageTypeStr := parts[0]
 	pkg = parts[1]
-	ptParts := strings.Split(packageTypeStr, ".")
+	ptParts := strings.Split(packageTypeStr, "#")
```
Comment (apelisse) [inline staging/src/k8s.io/code-generator/cmd/applyconfiguration-gen/args/externaltypes.go:117]: Below the comment reads as if `#` was expected?

**Comment by apelisse** [comment]
Why isn't this breaking any test? Would you mind adding a test to show that this works?

File: staging/src/k8s.io/code-generator/cmd/applyconfiguration-gen/args/externaltypes.go (hunk 2 above)
Comment (jpbetz) [reply staging/src/k8s.io/code-generator/cmd/applyconfiguration-gen/args/externaltypes.go:117]: Correct, on line 119, also need to update line 113.

Comment (jpbetz) [reply staging/src/k8s.io/code-generator/cmd/applyconfiguration-gen/args/externaltypes.go:117]: Oh wait. To retain compatiblity for existing users,  split on the last `.` here? This which will work regardless of how many dots are in the group name.  `strings.LastIndex` can help with this.

**Comment by xiaoyawei** [comment]
@jpbetz @apelisse thanks for the feedback but this issue was fixed 3 months ago in https://github.com/kubernetes/kubernetes/pull/111934/files

File: staging/src/k8s.io/code-generator/cmd/applyconfiguration-gen/args/externaltypes.go (hunk 2 above)
Comment (apelisse) [reply staging/src/k8s.io/code-generator/cmd/applyconfiguration-gen/args/externaltypes.go:117]: I guess the comment below, line 119, still needs to be fixed if someone wants to make a super easy PR.

**Pull request closed without merging**